specs/kiota.config.md

```diff
@@ -0,0 +1,634 @@
+# Kiota Config
+
+Kiota generates client code for an API and stores parameters in a kiota.lock file. A project can contain multiple API clients, but they are independently managed. Kiota has no awareness that an app has a dependency on multiple APIs, even though that is a core use case.
+
+## Status
+
+| Date | Version | Author | Status |
+| -- | -- | -- | -- |
+| November 30th, 2023 | v0.3 | Sébastien Levert | Final Draft |
+| November 22nd, 2023 | v0.2 | Sébastien Levert | Draft |
+| September 24th, 2023 | v0.1 | Darrel Miller | Draft |
+
+## Current Challenges
+
+- Client code generation is not reproducible if API description changes
+- Kiota doesn’t have a good solution for APIs that use multiple security schemes.
+- Kiota doesn’t provide any support for generating auth providers with the required permissions, partially because currently we generate one client for APIs that use different schemes. How would we know which auth provider to generate.
+- Kiota doesn’t have a good story for acquiring a client identifier. e.g. apikey or OAuth2 ClientId. This could be possible if the OpenIDConnect URL pointed to a dynamic registration endpoint.
+- If an application has multiple kiota clients, there is currently no way perform operations that correspond to all of the clients.
+
+We have previously described Kiota's approach to managing API dependencies as consistent with the way people manage packages in a project. However, currently our tooling doesn't behave that way. We treat each dependency independently.
+
+## Proposal
+
+We should introduce a new Kiota.config file that holds the input parameters required to generate the API Client code. Currently kiota.lock is used to capture what the parameters were at the time of generation and can be used to regenerate based on the parameters in the file. This creates a mixture of purposes for the file.
+
+We did consider creating one kiota.config file as as a peer of the language project file, however, for someone who wants to generate multiple clients for an API in different languages, this would be a bit annoying. An alternative would be to allow the kiota.config file to move further up the folder structure and support generation in multiple languages from a single file. This is more consistent with what [TypeSpec](https://aka.ms/typespec) are doing and would be helpful for generating CLI and docs as well as a library.
+
+Here is an example of what the kiota.config file could look like.
+
+```json
+{
+  "name": "My application",
+  "apis": {
+    "Graph": {
+      "descriptionHash": "9EDF8506CB74FE44...",
+      "descriptionLocation": "https://.../openapi.yaml",
```

Comment:
we might want to include the search key, especially when we talk about integrating with multiple registries and assuming the location under a search key can change.
I think the location should only be included in the config file when it's been provided by the user directly.
Associated to that. the "resolved location" (where did we put the local copy) should be in the lock file.

Reply:
Why? Once it's "found", I expect the absolute location to be the only thing we need. What would be the use case here?

Reply:
search key remains the same but the URI of the description changes between initial generation and update